File: go/src/recursion/quick_sort.go
package recursion

func QuickSort(arr []int) []int {
	if len(arr) <= 1 {
		return arr
	}

	pivot := arr[len(arr)-1]

	left := []int{}
	right := []int{}

	for i := 0; i < len(arr)-1; i++ {
		if arr[i] < pivot {
			left = append(left, arr[i])
		} else {
			right = append(right, arr[i])
		}
	}

	left = QuickSort(left)
	right = QuickSort(right)

	left = append(left, pivot)

	return append(left, right...)
}

File: rust/src/recursion/quick_sort.rs
pub fn quick_sort(arr: Vec<i32>) -> Vec<i32> {
    if arr.len() <= 1 { return arr; }

    let pivot = arr[arr.len()-1];

    let mut left: Vec<i32> = Vec::new();
    let mut right: Vec<i32> = Vec::new();

    arr[..arr.len()-1].iter().for_each(|&num| {
        if num <= pivot {
            left.push(num);
        } else {
            right.push(num);
        }
    });
    
    left = quick_sort(left);
    right = quick_sort(right);

    left.push(pivot);
    left.append(&mut right);

    return left;
}
